pkg/download/downloader_test.go

```diff
@@ -37,24 +37,91 @@ func testdataPath() string {
 	return filepath.Join(pwd, "testdata")
 }
 
+type ArchiveTest struct {
+	in      string
+	files   []string
+	wantErr bool
+}
+
+func SymlinkTest1(ext string) ArchiveTest {
+	return ArchiveTest{
+		in: "test-with-symlinks" + ext,
+		files: []string{
+			"/symlinks/",
+			"/symlinks/message",
+			"/symlinks/msg",
+		},
+		wantErr: false,
+	}
+}
+
+func SymlinkTest2(ext string) ArchiveTest {
+	return ArchiveTest{
+		in: "test-with-symlinks-escaping-parent" + ext,
+		files: []string{
+			"/escaping-link-test2/",
+			"/escaping-link-test2/yo/",
+			"/escaping-link-test2/yo/mo/",
+			"/escaping-link-test2/yo/mo/zoo",
+			"/escaping-link-test2/z",
+			"/escaping-link-test2/zz",
+			"/escaping-link-test2/zzz", // this escapes only to the staging area, so wantErr: false
+		},
+		wantErr: false,
+	}
+}
+
+func SymlinkTest3(ext string) ArchiveTest {
+	return ArchiveTest{
+		in: "test-with-symlinks-escaping-parent2" + ext,
+		files: []string{
+			"/escaping-link-test3/",
+			"/escaping-link-test3/baz", // this escapes only to the staging area, so wantErr: false
+			"/escaping-link-test3/foo/",
+			"/escaping-link-test3/foo/bar/",
+		},
+		wantErr: false,
+	}
+}
+
+func SymlinkTest4(ext string) ArchiveTest {
+	return ArchiveTest{
+		in:      "test-with-symlinks-escaping-parent3" + ext,
+		files:   nil,
+		wantErr: true,
+	}
+}
+
+func SymlinkTest5(ext string) ArchiveTest {
```

Comment:
I think we have waaay to many tests here.

For testdata, we just need 1 symlink .tar.gz and 1 .zip in total to verify that works.

For everything else (such as disallowed links), we can test them by unit-testing the method that decides it.

Reply:
sure. personally, i'm a fan of end-to-end tests when they are cheap.